### PR TITLE
Add keep-in constraints

### PR DESCRIPTION
**Pull request by Lewis-Seiden**

### CODE REVIEW

File: src/components/field/svg/constraintDisplay/KeepInCircleOverlay.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    return (
+      <g ref={this.rootRef}>
+        {/* Main Circle */}
+        <circle
+          cx={x}
+          cy={y}
+          r={r - STROKE / 2}
+          fill={"green"}
+          fillOpacity={0.1}
+          id="dragTarget-keepInCircle"
```
Comment (shueja) [inline src/components/field/svg/constraintDisplay/KeepInCircleOverlay.tsx:77]: These IDs are going to break if there's ever multiple overlays active at once. These need to be globally unique in the document

File: src/components/field/svg/constraintDisplay/KeepInCircleOverlay.tsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        .on("end", (event) => doc.history.stopGroup())
+        .container(this.rootRef.current);
+      d3.select<SVGCircleElement, undefined>(`#dragRadiusTarget-keepInCircle`).call(
+        radiusHandleDrag
+      );
+    }
+  }
+
+  dragPointTranslate(event: any) {
+    this.props.data.x.set(this.props.data.serialize.props.x[1] + event.dx);
```
Comment (shueja) [inline src/components/field/svg/constraintDisplay/KeepInCircleOverlay.tsx:51]: you can do `this.props.data.props.x.value` instead

File: src/components/field/svg/constraintDisplay/KeepInRectangleOverlay.tsx [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+          doc.history.stopGroup();
+        })
+        .container(this.rootRef.current);
+      d3.select<SVGCircleElement, undefined>(`#dragTarget-keepInRectangleRegion`).call(
+        dragHandleRegion
+      );
+    }
+  }
+
+  dragPointTranslate(event: any, xOffset: boolean, yOffset: boolean) {
```
Comment (shueja) [inline src/components/field/svg/constraintDisplay/KeepInRectangleOverlay.tsx:103]:
```suggestion
  dragPointTranslate(event: any, xOffset: boolean, yOffset: boolean) {
  const data = this.props.data;
  //saves retyping it every time
```

**Review by shueja — CHANGES_REQUESTED** [review]
This is going to merge conflict with @oh-yes-0-fps rework of generation logic. Please hold off on merge until then.

File: trajoptlib/include/trajopt/constraint/PointLineRegionConstraint.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    //   b = x₁ − x₀
+    //   c = (y₀ − y₁)x₀ + (x₁ − x₀)y₀
+    //     = ax₀ + by₀
+
+    // ax + by = ax₀ + by₀ where
+    //   a = y₀ − y₁
+    //   b = x₁ − x₀
+
+    problem.SubjectTo(
+      ((m_fieldLineStart.Y() - m_fieldLineEnd.Y()) * bumperCorner.X()) +
```
Comment (calcmogul) [inline trajoptlib/include/trajopt/constraint/PointLineRegionConstraint.hpp:74]: Some helper variables that match the derivation would make this shorter and easier to read.

**Review by shueja — CHANGES_REQUESTED** [review]
You're missing the RustFFI.cpp impl of DifferentialPathBuilder::sgmt_keep_in_polygon, which is causing link failures.

File: src-core/src/generation/generate.rs [hunk 5]
```diff
@@ -319,9 +326,20 @@ pub fn generate(
                     traj_builder.wpt_angular_velocity_max_magnitude(from, 0.0f64);
                 }
             }
+            ConstraintData::KeepInCircle { x, y, r } => match to_opt {
```
Comment (oh-yes-0-fps) [inline src-core/src/generation/generate.rs:329]: To merge you have to move this into constraints.rs transformer

File: src/components/field/svg/constraintDisplay/KeepInCircleOverlay.tsx (hunk 1 above)
Comment (Lewis-Seiden) [reply src/components/field/svg/constraintDisplay/KeepInCircleOverlay.tsx:77]: isnt this also a problem with the point at overlay?

**Review by shueja — APPROVED** [review]

**Review by calcmogul — COMMENTED** [review]
Some API docs are still missing.

File: trajoptlib/include/trajopt/constraint/PointLineRegionConstraint.hpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    // ax + by = ax₀ + by₀ where
+    //   a = y₀ − y₁
+    //   b = x₁ − x₀
+
+    auto a = m_fieldLineStart.Y() - m_fieldLineEnd.Y();
+    auto b = m_fieldLineEnd.X() - m_fieldLineStart.X();
+
+    problem.SubjectTo((a * bumperCorner.X()) + (b * bumperCorner.Y()) >
+                      (a * m_fieldLineStart.X()) + (b * m_fieldLineStart.Y()));
```
Comment (calcmogul) [inline trajoptlib/include/trajopt/constraint/PointLineRegionConstraint.hpp:79]:
```suggestion
    problem.SubjectTo(a * bumperCorner.X() + b * bumperCorner.Y() >
                      a * m_fieldLineStart.X() + b * m_fieldLineStart.Y());
```

File: trajoptlib/include/trajopt/constraint/PointLineRegionConstraint.hpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+#include "trajopt/constraint/detail/LinePointDistance.hpp"
+#include "trajopt/geometry/Pose2.hpp"
+#include "trajopt/geometry/Translation2.hpp"
+#include "trajopt/util/SymbolExports.hpp"
+
+namespace trajopt {
+
+/**
+ *
```
Comment (calcmogul) [inline trajoptlib/include/trajopt/constraint/PointLineRegionConstraint.hpp:19]: Needs a class-level description like the other constraints.

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: trajoptlib/include/trajopt/constraint/PointLineRegionConstraint.hpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+#include "trajopt/constraint/detail/LinePointDistance.hpp"
+#include "trajopt/geometry/Pose2.hpp"
+#include "trajopt/geometry/Translation2.hpp"
+#include "trajopt/util/SymbolExports.hpp"
+
+namespace trajopt {
+
+/**
+ * Point-LineRegion constraint.
```
Comment (calcmogul) [inline trajoptlib/include/trajopt/constraint/PointLineRegionConstraint.hpp:19]:
```suggestion
 * Point-Line region constraint.
```

**Review by calcmogul — APPROVED** [review]

**Pull request merged**